components/body_sniffer/body_sniffer_url_loader.cc

```diff
@@ -0,0 +1,282 @@
+    mojo::ScopedDataPipeConsumerHandle body) {
+  VLOG(2) << __func__ << " " << response_url_;
+  state_ = State::kLoading;
+  body_consumer_handle_ = std::move(body);
+  body_consumer_watcher_.Watch(
+      body_consumer_handle_.get(),
+      MOJO_HANDLE_SIGNAL_READABLE | MOJO_HANDLE_SIGNAL_PEER_CLOSED,
+      base::BindRepeating(&BodySnifferURLLoader::OnBodyReadable,
+                          base::Unretained(this)));
+  body_consumer_watcher_.ArmOrNotify();
+}
+
+void BodySnifferURLLoader::OnComplete(
+    const network::URLLoaderCompletionStatus& status) {
+  DCHECK(!complete_status_.has_value());
+  switch (state_) {
+    case State::kWaitForBody:
+      // An error occured before receiving any data.
+      DCHECK_NE(net::OK, status.error_code);
+      state_ = State::kCompleted;
+      if (!throttle_) {
+        Abort();
+        return;
+      }
+      throttle_->Resume();
+      destination_url_loader_client_->OnComplete(status);
+      return;
+    case State::kLoading:
+    case State::kSending:
+      complete_status_ = status;
+      return;
+    case State::kCompleted:
+      destination_url_loader_client_->OnComplete(status);
+      return;
+    case State::kAborted:
+      NOTREACHED();
+      return;
+  }
+  NOTREACHED();
+}
+
+void BodySnifferURLLoader::FollowRedirect(
+    const std::vector<std::string>& removed_headers,
+    const net::HttpRequestHeaders& modified_headers,
+    const net::HttpRequestHeaders& modified_cors_exempt_headers,
+    const absl::optional<GURL>& new_url) {
+  // BodySnifferURLLoader starts handling the request after
+  // OnReceivedResponse(). A redirect response is not expected.
+  NOTREACHED();
+  source_url_loader_->FollowRedirect(removed_headers, modified_headers,
+                                     modified_cors_exempt_headers, new_url);
+}
+
+void BodySnifferURLLoader::SetPriority(net::RequestPriority priority,
+                                       int32_t intra_priority_value) {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->SetPriority(priority, intra_priority_value);
+}
+
+void BodySnifferURLLoader::PauseReadingBodyFromNet() {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->PauseReadingBodyFromNet();
+}
+
+void BodySnifferURLLoader::ResumeReadingBodyFromNet() {
+  if (state_ == State::kAborted)
+    return;
+  source_url_loader_->ResumeReadingBodyFromNet();
+}
+
+bool BodySnifferURLLoader::CheckBufferedBody(uint32_t readBufferSize) {
+  size_t start_size = buffered_body_.size();
+  uint32_t read_bytes = readBufferSize;
+  buffered_body_.resize(start_size + read_bytes);
+
+  auto result = body_consumer_handle_->ReadData(
+      &buffered_body_[0] + start_size, &read_bytes, MOJO_READ_DATA_FLAG_NONE);
+  switch (result) {
+    case MOJO_RESULT_OK:
+      return true;
+    case MOJO_RESULT_FAILED_PRECONDITION:
+      buffered_body_.resize(start_size);
+      CompleteLoading(std::move(buffered_body_));
```

Comment:
IMHO moving class member is a bad practice. Also we always call CompleteLoading with `buffered_body_`. Could it in CompleteLoading() directly or via a getter?